plugin/core/protocol.py

```diff
@@ -364,6 +364,37 @@ class SemanticTokenModifiers:
     'watchers': List[FileSystemWatcher],
 }, total=True)
 
+InlayHintParams = TypedDict('InlayHintParams', {
+    'textDocument': TextDocumentIdentifier,
+    'range': RangeLsp,
+}, total=True)
+
+InlayHintLabelPart = TypedDict('InlayHintLabelPart', {
+    'value': str,
+    'tooltip': Union[str, MarkupContent],
+    'location': Location,
+    'command':  Command
```

Comment:
I don't really agree with this. Using typing's `Optional` would mean for the type checker that the value could be `None`, while it would be a spec violation if the server would actually send a `null` JSON value. If desired for documentation, we could mark optional properties with a comment.

And imo the TypedDicts that currently are declared with `total=True` and use `Optional` for optional properties should be converted to `total=False` with `Optional` removed to make the file more consistent.

Using `Optional` here would also be confusing regarding LSP types that have properties explicitly allowing the value `null` with a special meaning, e.g. [WorkspaceFullDocumentDiagnosticReport](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#workspaceFullDocumentDiagnosticReport).

Reply:
Maybe we can backport `typing.Required` and `typing.NotRequired` from Python 3.11 ([PEP 655](https://peps.python.org/pep-0655/#specification))? That would probably be the best solution for the pyright/mypy type checkers.

Reply:
As we discussed before, this is a choice between two wrong solutions since python typing doesn't allow us to express this correctly.

Now the question is which one is "less wrong".

- with `total=True` we can access required properties without "null" checking (positive) while with `total=False` we have to null check all (negative).
- with `total=True` directly accessing optional property won't raise a type error (negative)
- `total=True` is problematic for objects that we have to construct as then we need to specify all properties, even optional ones (negative)

I guess with that list I might have convinced myself that `total=True` is worse... :)
Though with `total=False` we are often creating Pyright type errors (or silence them with `type: ignore` which is something I don't like.

Reply:
With that said, you should choose what you prefer @predragnikolic. I will stay silent on those issues until proper python solution is out.

Reply:
> I guess with that list I might have convinced myself that `total=True` is worse... :) 

:)

> Though with `total=False` we are often creating Pyright type errors (or silence them with `type: ignore`) which is something I don't like.

For that reason I currently use
```
"python.analysis.diagnosticSeverityOverrides": {
    "reportTypedDictNotRequiredAccess": "none",
}
```
in my LSP-pyright settings.

I will try to add a backport for `typing.NotRequired` to see if the Pyright errors go away with it, and open a PR if it works.

Reply:
I guess should add to the negative that without `Optional`, even if "more correct", when reading the code I have no idea if given property is require or not and it's hard for me to reason about the flow of the logic. I need to lookup actual types in LSP spec for clarification.

Reply:
I hope that in the following 2 months we will migrate to py 3.8.
and include the typing_extension package in ST and properly use the `NotRequired` type.